### PR TITLE
feat: thea pallet and improvements to asset-handler pallet

### PR DESCRIPTION
**Pull request by felixfaisal**
## Checklist before requesting a review
- [x] I have performed a self-review of my code.
- [x] If it is a core feature, I have added thorough tests.
- [x] I removed all Clippy and Formatting Warnings. 
- [x] I added required Copyrights.

### CODE REVIEW

**Comment by felixfaisal** [comment]
Hey @serhii-temchenko The latest CI build has an error related to dependencies 
```bash
error[E0658]: use of unstable library feature 'scoped_threads'
``` 
Can you take a look into this?

**Comment by serhii-temchenko** [comment]
> Hey @serhii-temchenko The latest CI build has an error related to dependencies
> 
> ```shell
> error[E0658]: use of unstable library feature 'scoped_threads'
> ```
> 
> Can you take a look into this?

Sure, error appeared in the process of formatter installation. Required version (1.5) was outdated and for now in latest rustup installer present 1.5.1 version. So we don't need install formatter separately. I raised the PR #598 with fix of this and some refactoring of workflows declaration. Please review that PR and after merge it - you could pick fix from the `Develop` branch.

File: pallets/asset-handler/src/lib.rs
Comment (serhii-temchenko) [inline pallets/asset-handler/src/lib.rs]: Missed underscore in `identifier length`

**Review by serhii-temchenko — CHANGES_REQUESTED** [review]

File: pallets/asset-handler/src/lib.rs
Comment (felixfaisal) [reply pallets/asset-handler/src/lib.rs]: Fixed ✅

**Review by serhii-temchenko — DISMISSED** [review]

File: pallets/asset-handler/src/lib.rs
Comment (Gauthamastro) [inline pallets/asset-handler/src/lib.rs:177]: Documentation on this

File: pallets/asset-handler/src/lib.rs
Comment (Gauthamastro) [inline pallets/asset-handler/src/lib.rs:343]: Please add a check to ensure indexing won't panic

File: pallets/asset-handler/src/tests.rs
Comment (Gauthamastro) [inline pallets/asset-handler/src/tests.rs:590]: Add a case where the recipient account doesn't exist and Thea mints into it. or burns. Let's see the error handling

**Review by Gauthamastro — CHANGES_REQUESTED** [review]

File: pallets/asset-handler/src/lib.rs
Comment (felixfaisal) [reply pallets/asset-handler/src/lib.rs:177]: Done ✅

File: pallets/asset-handler/src/lib.rs
Comment (felixfaisal) [reply pallets/asset-handler/src/lib.rs:343]: Done ✅

File: pallets/asset-handler/src/tests.rs
Comment (felixfaisal) [reply pallets/asset-handler/src/tests.rs:590]: Done ✅

**Review by Gauthamastro — DISMISSED** [review]

**Comment by serhii-temchenko** [comment]
@felixfaisal please check PR #602.

File: node/src/service.rs
Comment (serhii-temchenko) [inline node/src/service.rs]: Please remove

File: node/src/service.rs
Comment (serhii-temchenko) [inline node/src/service.rs]: Please remove

File: pallets/asset-handler/src/lib.rs
Comment (serhii-temchenko) [inline pallets/asset-handler/src/lib.rs]: Please use 3 slashes as above.

File: pallets/thea/Cargo.toml
Comment (serhii-temchenko) [inline pallets/thea/Cargo.toml]: Please remove extra empty line

File: pallets/thea/src/lib.rs
Comment (serhii-temchenko) [inline pallets/thea/src/lib.rs]: Please remove if not needed

File: pallets/thea/src/lib.rs
Comment (serhii-temchenko) [inline pallets/thea/src/lib.rs:155]: Please add documentation as well

File: pallets/thea/src/lib.rs
Comment (serhii-temchenko) [inline pallets/thea/src/lib.rs]: Is it ok that we are popping account on line 227 and if no approved deposits for this account yet - we just removing it from storage by overwriting here on line 251? shouldn't we put it back in accounts collection if there's no approved deposits yet? if no - at least some kind of log should be in `else` case between 248 and 249 lines.

File: pallets/thea/src/lib.rs
Comment (serhii-temchenko) [inline pallets/thea/src/lib.rs:254]: Is there a task for the benchmark?

File: pallets/thea/src/lib.rs
Comment (serhii-temchenko) [inline pallets/thea/src/lib.rs:259]: Dockblock has a different set of args. Please update.

File: pallets/thea/src/lib.rs
Comment (serhii-temchenko) [inline pallets/thea/src/lib.rs:265]: I suggest to move this bloc after line 303 since its required on line 306 only.

File: pallets/thea/src/lib.rs
Comment (serhii-temchenko) [inline pallets/thea/src/lib.rs]: Unhandled unwrap.

File: pallets/thea/src/lib.rs
Comment (serhii-temchenko) [inline pallets/thea/src/lib.rs:299]: Unhandled unwrap

File: pallets/thea/src/lib.rs
Comment (serhii-temchenko) [inline pallets/thea/src/lib.rs:298]: If it's not - shouldn't be created vector and still pushed approved deposit?

File: thea-primitives/Cargo.toml
Comment (serhii-temchenko) [inline thea-primitives/Cargo.toml]: Please add space after comma `"polkadot-v0.9.28",default-features`

File: thea-primitives/Cargo.toml
Comment (serhii-temchenko) [inline thea-primitives/Cargo.toml]: Please add empty line

File: pallets/thea/src/lib.rs
Comment (serhii-temchenko) [inline pallets/thea/src/lib.rs:338]: Is it proper logic to do nothing, when calling this extrinsic, if no approved deposits? mby some proper error code should be returned of at least logged? or it should be idempotent?

File: pallets/thea/src/lib.rs
Comment (serhii-temchenko) [inline pallets/thea/src/lib.rs:364]: Is it possible that by time of execution of loop from above some new elements could appeared in storage (ApprovedDeposits)? so we will end up with hanging ApprovedDeposits since account was removed from the pending for processing list?

File: pallets/thea/src/lib.rs
Comment (serhii-temchenko) [inline pallets/thea/src/lib.rs:357]: Shouldn't this deposit (if no errors on calling `execute_deposit`) be removed from the storage (ApprovedDeposits) since transaction was made to not make it permanent?

File: pallets/thea/src/lib.rs
Comment (serhii-temchenko) [inline pallets/thea/src/lib.rs]: When this should be implemented? please add in-code reference to the issue

File: pallets/thea/src/lib.rs
Comment (serhii-temchenko) [inline pallets/thea/src/lib.rs:419]: Dockblock has different amount of parameters. Please update.

File: pallets/thea/src/lib.rs
Comment (serhii-temchenko) [inline pallets/thea/src/lib.rs]: Please elaborate on this comment

File: pallets/thea/src/lib.rs
Comment (serhii-temchenko) [inline pallets/thea/src/lib.rs]: Same

File: pallets/thea/src/lib.rs
Comment (serhii-temchenko) [inline pallets/thea/src/lib.rs]: Please add some task reference to not forget

**Review by serhii-temchenko — CHANGES_REQUESTED** [review]
Only 3 assertions present in a tests for `approve_deposit` extrinsic. Please add all required tests and assertions for the `thea` pallet.

File: pallets/thea/src/lib.rs
Comment (felixfaisal) [reply pallets/thea/src/lib.rs:265]: I don't get this comment, can you please elaborate

File: pallets/thea/src/lib.rs
Comment (felixfaisal) [reply pallets/thea/src/lib.rs:299]: This will be refactored, Will add a comment

File: pallets/thea/src/lib.rs
Comment (serhii-temchenko) [reply pallets/thea/src/lib.rs:265]: Sure, this selected block could be moved lower by execution chain since `nonce` required only on line number 306. So by moving 281-283 lines after 303 - other assertions would be executed first and calculation time in invalid scenario will be faster.

File: pallets/thea/src/lib.rs
Comment (felixfaisal) [reply pallets/thea/src/lib.rs:254]: Yes, I've created an issue for it

File: pallets/thea/src/lib.rs
Comment (felixfaisal) [reply pallets/thea/src/lib.rs:155]: Added ✅

File: pallets/thea/src/lib.rs
Comment (felixfaisal) [reply pallets/thea/src/lib.rs]: Remove ✅

File: pallets/thea/Cargo.toml
Comment (felixfaisal) [reply pallets/thea/Cargo.toml]: Remove ✅

File: pallets/asset-handler/src/lib.rs
Comment (felixfaisal) [reply pallets/asset-handler/src/lib.rs]: Added ✅

File: node/src/service.rs
Comment (felixfaisal) [reply node/src/service.rs]: Removed ✅

File: node/src/service.rs
Comment (felixfaisal) [reply node/src/service.rs]: Removed ✅

File: pallets/thea/src/lib.rs
Comment (felixfaisal) [reply pallets/thea/src/lib.rs:259]: Updated ✅

File: pallets/thea/src/lib.rs
Comment (felixfaisal) [reply pallets/thea/src/lib.rs]: Removed unwrap ✅

File: pallets/thea/src/lib.rs
Comment (felixfaisal) [reply pallets/thea/src/lib.rs:298]: The scenario is not possible, if `Storage item` contains the key

File: thea-primitives/Cargo.toml
Comment (felixfaisal) [reply thea-primitives/Cargo.toml]: Updated ✅

File: thea-primitives/Cargo.toml
Comment (felixfaisal) [reply thea-primitives/Cargo.toml]: Added ✅

File: pallets/thea/src/lib.rs
Comment (felixfaisal) [reply pallets/thea/src/lib.rs:338]: Added an else case, which returns an error

File: pallets/thea/src/lib.rs
Comment (felixfaisal) [reply pallets/thea/src/lib.rs:364]: This race condition should not be possible, It will not be executed in parallel, The `on_idle` hook will only make use of any remaining weight if available, that is when there is no activity in the blockchain. 
CC: @zktony @Gauthamastro

File: pallets/thea/src/lib.rs
Comment (felixfaisal) [reply pallets/thea/src/lib.rs:357]: We are beginning this code block with `deposits.pop()` so if it does not return any error, It's not in the storage item, and then after the loop we are overwriting the `Storage Item`, `Approved Deposits` with the `deposits` vector. So it is being removed.

File: pallets/thea/src/lib.rs
Comment (felixfaisal) [reply pallets/thea/src/lib.rs]: This will be implemented when work on withdrawal begins

File: pallets/thea/src/lib.rs
Comment (felixfaisal) [reply pallets/thea/src/lib.rs:419]: Updated ✅

File: pallets/thea/src/lib.rs
Comment (felixfaisal) [reply pallets/thea/src/lib.rs]: This will be refactored when work on withdrawal so not fixing clippy suggestion for now

File: pallets/thea/src/lib.rs
Comment (felixfaisal) [reply pallets/thea/src/lib.rs:265]: Other assertions are computationally more expensive, So isn't it better to have it here?

File: pallets/thea/src/lib.rs
Comment (felixfaisal) [reply pallets/thea/src/lib.rs]: Done

File: pallets/thea/src/lib.rs
Comment (Gauthamastro) [reply pallets/thea/src/lib.rs:364]: Yes, on_idle will execute only when there are no more extrinsic to include in the block and enough remaining weight to complete the approve_deposit.

File: pallets/thea/src/lib.rs
Comment (serhii-temchenko) [reply pallets/thea/src/lib.rs:265]: Ok

File: pallets/thea/src/lib.rs
Comment (serhii-temchenko) [reply pallets/thea/src/lib.rs:299]: Still unresolved

**Comment by serhii-temchenko** [comment]
@felixfaisal still 2 comments untouched. Please check them. 
Plus nothing about [this](https://github.com/Polkadex-Substrate/Polkadex/pull/595#pullrequestreview-1246288395) comment.

File: pallets/thea/src/lib.rs
Comment (felixfaisal) [reply pallets/thea/src/lib.rs:299]: Added a comment

File: pallets/thea/src/lib.rs
Comment (felixfaisal) [reply pallets/thea/src/lib.rs]: It's fine, All the accounts are accounts that have pending deposits, which is being called here 
```rust
let mut accounts = <AccountWithPendingDeposits<T>>::get();
```
CC: @Gauthamastro

**Comment by felixfaisal** [comment]
Hey @serhii-temchenko Tests will be covered under a separate issue - https://github.com/Polkadex-Substrate/Polkadex/issues/614

**Review by serhii-temchenko — APPROVED** [review]

**Review by Gauthamastro — APPROVED** [review]

**Pull request merged**
